_posts/2021-04-14-removing-implicit-coercions.markdown

```diff
@@ -0,0 +1,215 @@
+layout: post
+author: dizzy
+category: blog
+---
+
+tl/dr: we’re doubling down on removing surprising implicit (and some explicit) coercions from the language. So far we
+
+* removed object → num conversion (it now throws an exception)
+* have pieces in flight for the removal of coercions due to string interpolation/concatenation, mathematical operations (including bitwise), and pre/post increment/decrement
+* have concrete plans for comparison operators, the (in)equality operators, and switch statements.
+
+## Context
+
+Implicit coercions can be unintuitive in edge cases, and these edge cases can be surprisingly common - making them the second most confusing language feature, behind arrays. Even those working on the language often make mistakes, and they are frequently a source of subtle bugs.
+
+For those who don’t know, an implicit coercion is when a value of type A gets converted to type B as part of computing the result of an operation. Probably the most common example is how `0 == false` results in `true` courtesy of `0` being converted to `false` prior to the comparison, but things go straight downhill from there (hint: try comparing `'0'` with `false` and then `'0'` with `null`). We’ve determined that the time has finally come to tackle this once and for all. 
+
+## So what are we going to do about it?
+
+Well, we’ve actually already started doing something about it! Very shortly, attempting to convert an object into a number will universally throw an exception. On top of this almost finished step, we have a few portions in flight, as well as a few more not started but still very much planned.
+
+For the following, the possible stages are
+
+1. No Progress - Allowed by both the Type Checker and runtime. Lints may exist.
+2. Enforced By Type Checker - the Type checker will report errors
+3. Soft Runtime Enforcement - Violations may trigger logs via the runtime, configurable via flag.
+4. Hard Runtime Enforcement - Violations will produce a runtime exception
+5. New Behaviour - Following stage 4, we may decide to re-allow the operation but with a different result.
+
+These stages are ordered such that by a certain stage, all previous stages are implied.
+
+### String Concatenation (`.` , `.=`) interpolation
+
+**Current Behaviour:**
+int, null, bool, float, and resource will implicitly coerce to string
+
+```
+'' . true; // '1'
+```
+
+**Future Behaviour:**
+*only* int will implicitly coerce. The rest will trigger exceptions
+Example:
+
+```
+'' . true; // `InvalidOperationException` 
+'one: ' . 1; // 'one: 1'
+```
+
+**Status of rollout:** 
+Soft Runtime Enforcement
+
+### Bitwise operations (`&`, `|`, `^`, `~`, `<<`, `>>`) and assignment equivalents
+
+**Current Behaviour:**
+
+`&`, `|`, `^`: null, bool, float, vec, dict, keyset, and resource will implicitly coerce to int. strings will implicitly coerce as well *unless* both operands are strings.
+
+```
+null & vec[]; // 0
+keyset[1] & true; // 1
+'abcd' & 'efgh'; // 'abc`'
```

Comment:
might be worth clarifying this one: ascii('d') & ascii('h') == ascii('`'), or link to https://3v4l.org/CTH2o